types/query-string/index.d.ts

```diff
@@ -13,13 +13,21 @@ export interface ParseOptions {
     decode?: boolean;
 }
 
+export interface InputParams {
+  [key: string]: any;
+}
+
+export interface OutputParams {
+  [key: string]: string | string[] | undefined;
```

Comment:
What is the use case for the `undefined` possibility? Is this to catch e.g. an empty query string? I'm not sure it's necessary to have that, but it may be worth adding a test case to validate this use case.

Reply:
This gets around what I feel like is an inconsistency in TypeScript. Here's an illustration:

The object `{a: 'hi'}` fits into the type `{[key: string}: string}`. However that type represents "an object whose values at all string indices is a string". This is not true -- only the value at `'a'` is a string. The returned value at all other indices is `undefined`.

This is something that TypeScript thinks is correct but will error at runtime:

```ts
let obj: {[key: string}: string} = {a: 'hi'};
obj['b'].length // Uncaught TypeError: Cannot read property 'length' of undefined
```

If we instead used the type `{[key: string}: string | undefined}`, we will get a static error for the above.

Reply:
Here's a more relevant example:

```ts
import queryString from "query-string";
let params = queryString.parse("a=1");
params['b'].length
```

This will error at runtime without `undefined`. It will be a static type error with `undefined`.

Reply:
Interesting. Makes sense to me, thanks.